File: src/DataScience/R/patterns/interpreter.R
# Interpreter: evaluate a tiny AST against a context.
env<-list(x=4); eval_expr<-function(n) switch(n[[1]],lit=n[[2]],var=env[[n[[2]]]],add=eval_expr(n[[2]])+eval_expr(n[[3]])); stopifnot(eval_expr(list('add',list('var','x'),list('lit',3)))==7)

File: src/Scripting/PythonPY/patterns/interpreter.py
"""Interpreter pattern: evaluate a tiny expression language with Python data and recursion.

Grammar (AST form):
    Expr := ("lit", int) | ("var", name) | ("add", Expr, Expr)
"""
from __future__ import annotations

from typing import TypeAlias

Expr: TypeAlias = tuple


def interpret(node: Expr, context: dict[str, int]) -> int:
    kind, *args = node
    if kind == "lit":
        return int(args[0])
    if kind == "var":
        return context[str(args[0])]
    if kind == "add":
        return interpret(args[0], context) + interpret(args[1], context)
    raise ValueError(f"unsupported expression: {kind}")


def main() -> None:
    expression: Expr = ("add", ("var", "x"), ("lit", 3))
    value = interpret(expression, {"x": 4})
    assert value == 7
    print("interpreter=7")


if __name__ == "__main__":
    main()
